pkg/sdn/plugin/subnets.go

```diff
@@ -168,43 +168,58 @@ func (master *OsdnMaster) clearInitialNodeNetworkUnavailableCondition(node *kapi
 }
 
 func (master *OsdnMaster) watchNodes() {
-	nodeAddressMap := map[types.UID]string{}
-	RunEventQueue(master.kClient.Core().RESTClient(), Nodes, func(delta cache.Delta) error {
-		node := delta.Object.(*kapi.Node)
-		name := node.ObjectMeta.Name
-		uid := node.ObjectMeta.UID
-
-		nodeIP, err := getNodeIP(node)
-		if err != nil {
-			return fmt.Errorf("failed to get node IP for %s, skipping event: %v, node: %v", name, delta.Type, node)
-		}
+	nodeInformer := master.informers.InternalKubernetesInformers().Core().InternalVersion().Nodes()
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			node := obj.(*kapi.Node)
+			master.handleAddOrUpdateNode(node, watch.Added)
+		},
+		UpdateFunc: func(_, cur interface{}) {
+			node := cur.(*kapi.Node)
+			master.handleAddOrUpdateNode(node, watch.Modified)
+		},
+		DeleteFunc: func(obj interface{}) {
+			obj, err := getDeletedObjFromInformer(obj, Nodes)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			node := obj.(*kapi.Node)
+			master.handleDeleteNode(node)
+		},
+	})
+}
 
-		switch delta.Type {
-		case cache.Sync, cache.Added, cache.Updated:
-			master.clearInitialNodeNetworkUnavailableCondition(node)
+func (master *OsdnMaster) handleAddOrUpdateNode(node *kapi.Node, eventType watch.EventType) {
+	nodeIP, err := getNodeIP(node)
+	if err != nil {
+		log.Errorf("Failed to get node IP for node %s, skipping %s event, node: %v", node.Name, eventType, node)
+		return
+	}
+	master.clearInitialNodeNetworkUnavailableCondition(node)
 
-			if oldNodeIP, ok := nodeAddressMap[uid]; ok && ((nodeIP == oldNodeIP) || isValidNodeIP(node.Status.Addresses, oldNodeIP)) {
-				break
-			}
-			// Node status is frequently updated by kubelet, so log only if the above condition is not met
-			log.V(5).Infof("Watch %s event for Node %q", delta.Type, name)
+	if oldNodeIP, ok := master.hostSubnetNodeIPs[node.UID]; ok && ((nodeIP == oldNodeIP) || isValidNodeIP(node.Status.Addresses, oldNodeIP)) {
+		return
+	}
+	// Node status is frequently updated by kubelet, so log only if the above condition is not met
+	log.V(5).Infof("Watch %s event for Node %q", eventType, node.Name)
 
-			usedNodeIP, err := master.addNode(name, nodeIP, nil, node.Status.Addresses)
-			if err != nil {
-				return fmt.Errorf("error creating subnet for node %s, ip %s: %v", name, nodeIP, err)
-			}
-			nodeAddressMap[uid] = usedNodeIP
-		case cache.Deleted:
-			log.V(5).Infof("Watch %s event for Node %q", delta.Type, name)
-			delete(nodeAddressMap, uid)
+	usedNodeIP, err := master.addNode(node.Name, nodeIP, nil, node.Status.Addresses)
+	if err != nil {
+		log.Errorf("Error creating subnet for node %s, ip %s: %v", node.Name, nodeIP, err)
+		return
+	}
+	master.hostSubnetNodeIPs[node.UID] = usedNodeIP
```

Comment:
Yes, add/update/del events are serially processed from the shared informer queue. So no additional locking required.